Validate JSON schema of API representers

NobodysNightmare opened the pull request:
This PR introduces some basic building blocks to load and validate schemas and starts using it.

Those self-tests are "basic" in the sense that they only validate
their compliance with our documented schema in one representation.

These test cases don't yet cover/validate whether the generated
representation also fulfills the schema under different circumstances,
for example when rendering for a user with fewer privileges, not allowed
to see certain fields.

Where necessary, the schema was changed to reflect the reality, e.g.
when those tests revealed that a "required" field might be missing due to
a lack of permissions.

# Ticket

https://community.openproject.org/wp/69707

# Notes

In a few cases the implementation was adapted to allow for stricter guarantees
of the specified schema, for example:

* links are allowed to leave out the title key already, so its not necessary to emit `title: nil` in cases where a title is
not known.
* `avatars` were documented to be in url format, though missing avatars were indicated as an empty string
    * this has been changed to representing missing avatars as nil, functionality does not appear to be affected by this

NobodysNightmare committed:
Add RSpec matcher for JSON schema

Uses json_schemer for the validation part.
Adding a helper service to the app for loading schema
from the docs, which is also intended to be used from
the application code, e.g. to indicate the schema to clients.

Specifically this is planned to be used for MCP endpoints.

NobodysNightmare committed:
Self-test basic schema compliance for a few API models

Those self-tests are "basic" in the sense that they only validate
their compliance with our documented schema in one representation.

These test cases don't yet cover/validate whether the generated
representation also fulfills the schema under different circumstances,
for example when rendering for a user with fewer privileges, not allowed
to see certain fields.

Where necessary, the schema was changed to reflect the reality, e.g.
when those tests revealed that a "required" field might be missing due to
a lack of permissions.

In a few cases the implementation was adapted to allow for stricter guarantees
of the specified schema, for example links allowed to leave out the title key
already, so its not necessary to emit `title: nil` in cases where a title is
not known.

Kharonus commented on docs/api/apiv3/components/schemas/attachment_model.yml:
```diff
@@ -27,11 +24,19 @@ properties:
     description: The files MIME-Type as determined by the server
   digest:
-    type: string
+    type: [string, null]
```
🟡 for OpenAPIv3.1 the better syntax for an array of types is:

```yaml
type:
  - string
  - "null"
```

e.g. read [here](https://www.openapis.org/blog/2021/02/16/migrating-from-openapi-3-0-to-3-1-0) or check the examples [here](https://spec.openapis.org/oas/v3.1.0.html). in YAML the arrays mostly are written without `[]`.

NobodysNightmare replied:
I have to admit that I find the single-line syntax easier to read for "something or null". This is a YAML feature, not an OpenAPI feature.

However, I see a valid remark in my use of `null`, which is interpreted as a literal null value when I intended to put the string `"null"`. So I'd probably suggest going for full JSON-mode:

```suggestion
    type: ["string", "null"]
```

WDYT?

Kharonus replied:
yes, the problem with the literal is something you MUST fix. still, the flow syntax for arrays is not the default syntax.

Usually it is preferred to use the block syntax for arrays, e.g. because of
- easier readable for longer lists or list with objects
- easier to diff/merge with git
- it keeps the indented YAML style and thus provides an easier flow when reading YAMLs

the flow syntax is usually only used when for some reasons it is important to keep the YAML compact.

As the API spec is something that is usually parsed by machines, I prefer if I can see syntax errors in a quick glance over a file without reading every line in detail, and that is way easier with the block syntax.

NobodysNightmare replied:
will change

NobodysNightmare replied:
changed

Kharonus commented on docs/api/apiv3/components/schemas/attachment_model.yml:
```diff
@@ -27,11 +24,19 @@ properties:
+    - prepared
+    - scanned
+    - quarantined
+    - rescan
```
🟡 something off with the formatting here, can you check your editor for formatting? the "better" approach is to have an indention for yaml arrays - i.e.

```yaml
  enum:
    - uploaded
    - prepared
    - scanned
    - quarantined
    - rescan
```

NobodysNightmare replied:
My editor accepts both and I usually prefer the less indented version, i.e.

```
enum:
- 1
```

instead of

```
enum:
  - 1
```

I started using this version, because it keeps indentation in nested structures lower and happens to be the same format that a `YAML.dump` in Ruby will emit.

If you have a strong opinion, I can change it.

Kharonus replied:
interesting, I have a strong visual preference for the second, as it looks like a bug to me, that the indention of a nested element is on the same level as its parent. :D I was even surprised, that it still is valid YAML.

So, in the end we should agree on one formatting option for YAML. How to find an agreement here?

NobodysNightmare replied:
I can give in for these enums easily.

One case where I find the deeper indentation outrageouly wasteful is when you have arrays of objects. Those could look like this:

```yaml
sensors:
  - id: 1
    name: Temperature outside
    history:
      - 18 °C
      - 20 °C
```

Here the history is already indented 6 spaces deep. Compare that to:


```yaml
sensors:
- id: 1
  name: Temperature outside
  history:
  - 18 °C
  - 20 °C
```

Indented two deep and it looks neat and concise. I guess critique would be that you need to see the `-` in line 2, to understand that the remaining lines are not _keys_ of `sensors`, but keys of the first element of an array.

NobodysNightmare replied:
updated. fun fact: I just realized that I didn't add the `allOf` above, which also used the short indentation :D

Kharonus commented on modules/avatars/spec/helpers/avatar_helper_spec.rb:
🟢 this makes sense! But did you check, if nothing breaks when we return now less empty strings and more `nil` in the avatar components?

NobodysNightmare replied:
I couldn't find anything breaking. I clicked around a little bit, but more or less unguided.

Kharonus commented on spec/fixtures/files/test_schema.yml:
```diff
@@ -0,0 +1,15 @@
+properties:
+  foo:
+    type: string
+    enum:
```
🟡 check yaml formatter, see comment above.

NobodysNightmare replied:
see reply above ^^

Kharonus commented on spec/lib/api/v3/types/type_representer_spec.rb:
```diff
@@ -37,7 +37,7 @@
 
   context "generation" do
-    subject { representer.to_json }
+    subject(:generated) { representer.to_json }
```
❓ why adding the name to the subject? tests below are using now both `subject` AND `generation`. seems inconsistent.

NobodysNightmare replied:
To be fair that's kind of an old problem. Other tests were also inconsistent, even when they were using `subject(:generated)`. I wanted to consistently add the spec with similar naming everywhere, therefore I added the alias when it was missing.

Question: What is the inconsistency we want to eliminate: Using `subject` and `generated` at the same time inside a spec or that among two specs that test the same thing one uses `subject` and the other uses `subject(:generated)`? xD

Kharonus replied:
the first one was the one crossing my mind :D If you deem it not a big issue, I can live with that. ;)

NobodysNightmare replied:
> If you deem it not a big issue, I can live with that. ;)

then let me self-resolve this conversation :D

Kharonus reviewed (changes requested):
I'm a big fan of this PR. 🥳 

Just have a small fine tuning suggestions.

NobodysNightmare committed:
Improve formatting of schema

In YAML, null is interpreted as a literal null value,
but we intended to use it as the string "null".

For consistency the first element of the array was also formatted
in double quotes and the array is now also formatted across
multiple lines.

Kharonus reviewed (approved):
❤️
